### PR TITLE
[jaeger-v2] Refactor ElasticSearch/OpenSearch Storage Configurations

### LINKED ISSUE #6059: [jaeger-v2] Refactor ElasticSearch/OpenSearch Storage Configurations To Align With OTEL
### Requirement

In this issue, we want to align the configuration for elastic search and open search storage with OTEL (part of https://github.com/jaegertracing/jaeger/issues/5229)

### Problem

 - [x] All config fields are tagged with mapstructure such that they can be addressable from YAML via "good" names
 - [x] Configs implement Validate()
 - [x] Configs via YAML support the same defaults as in jaeger-v1
 - [x] Configs reuse standard elements of OTEL configs whenever possible, e.g. TLS, clientcfg, etc.
 - [x] Configuration looks "logical" and properly grouped, not flattened with long weird names
 - [x] We create a migration guide document with a table that lists v1 CLI flags and the corresponding v2 config fields

### PR DESCRIPTION
**Pull request by mahadzaryab1**
## Which problem is this PR solving?
- Part of #6059

## Description of the changes
- Continuation of https://github.com/jaegertracing/jaeger/pull/5790
- Refactors the configurations for ElasticSearch and OpenSearch to simplify the configuration by having more logical groupings

## How was this change tested?
- Unit Tests and E2E Integration Tests

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: pkg/es/config/config.go [hunk 1]
```diff
@@ -33,48 +33,58 @@ import (
...
+	BulkSize                 int            `mapstructure:"-"`
+	BulkWorkers              int            `mapstructure:"-"`
+	BulkActions              int            `mapstructure:"-"`
+	BulkFlushInterval        time.Duration  `mapstructure:"-"`
+	Indices                  Indices        `mapstructure:"indices"`
+	ServiceCacheTTL          time.Duration  `mapstructure:"service_cache_ttl"`
+	AdaptiveSamplingLookback time.Duration  `mapstructure:"-"`
+	Tags                     TagsAsFields   `mapstructure:"tags_as_fields"`
+	Enabled                  bool           `mapstructure:"-"`
+	TLS                      tlscfg.Options `mapstructure:"tls"`
```
Comment (mahadzaryab1) [inline pkg/es/config/config.go:81]: we need to migrate this OTEL's TLS configuration - I can do that in a follow-up PR

File: pkg/es/config/config.go [hunk 2]
```diff
@@ -33,48 +33,58 @@ import (
...
+	AllowTokenFromContext    bool           `mapstructure:"-"`
+	Sniffer                  bool           `mapstructure:"sniffer"` // https://github.com/olivere/elastic/wiki/Sniffing
+	SnifferTLSEnabled        bool           `mapstructure:"sniffer_tls_enabled"`
+	MaxDocCount              int            `mapstructure:"-"` // Defines maximum number of results to fetch from storage per query
+	MaxSpanAge               time.Duration  `mapstructure:"-"` // configures the maximum lookback on span reads
+	Timeout                  time.Duration  `mapstructure:"-"`
+	BulkSize                 int            `mapstructure:"-"`
+	BulkWorkers              int            `mapstructure:"-"`
+	BulkActions              int            `mapstructure:"-"`
+	BulkFlushInterval        time.Duration  `mapstructure:"-"`
```
Comment (mahadzaryab1) [inline pkg/es/config/config.go:75]: @yurishkuro do we need mapstructure tags for these?

File: pkg/es/config/config.go (hunk 1 above)
Comment (yurishkuro) [reply pkg/es/config/config.go:81]: add TODO

File: pkg/es/config/config.go (hunk 2 above)
Comment (yurishkuro) [reply pkg/es/config/config.go:75]: I'd assume so. Are they configurable via CLI flags?

Comment (mahadzaryab1) [reply pkg/es/config/config.go:75]: @yurishkuro looks like it - can do in a follow-up PR as well

**Comment by mahadzaryab1** [comment]
@yurishkuro regarding https://github.com/jaegertracing/jaeger/pull/5790#discussion_r1749293974, are you referring to https://github.com/jaegertracing/jaeger/blob/main/plugin/storage/es/factory.go#L196-L210?

**Comment by yurishkuro** [comment]
> regarding #5790 (comment), are you referring to https://github.com/jaegertracing/jaeger/blob/main/plugin/storage/es/factory.go#L196-L210?

yes

File: plugin/storage/es/spanstore/reader.go [hunk 3]
```diff
@@ -128,23 +122,23 @@ func NewSpanReader(p SpanReaderParams) *SpanReader {
 	if p.UseReadWriteAliases {
 		maxSpanAge = rolloverMaxSpanAge
 	}
+
+	p.SpanIndex.Prefix = p.SpanIndex.FullIndexName(spanIndexBaseName)
+	p.ServiceIndex.Prefix = p.ServiceIndex.FullIndexName(serviceIndexBaseName)
```
Comment (mahadzaryab1) [inline plugin/storage/es/spanstore/reader.go:127]: @yurishkuro regarding the above, it looks like we have both span index and service index when constructing the span reader here

Comment (yurishkuro) [reply plugin/storage/es/spanstore/reader.go:127]: Reader doesn't matter as much because service index is only used for GetServices call, which would never be dispatched to the archive storage. But the archive writer could try to write to both spans and services indices. I just checked and doesn't seem like that would happen: https://github.com/jaegertracing/jaeger/blob/f411b3c49ccf5fb54abe3a024e772db28e4fc5b2/plugin/storage/es/spanstore/writer.go#L108

Here the empty string is returned for the services index name, and with blank string service name won't be written: https://github.com/jaegertracing/jaeger/blob/f411b3c49ccf5fb54abe3a024e772db28e4fc5b2/plugin/storage/es/spanstore/writer.go#L131

So I think the config in the original PR would still work, but one other thing I want to check is how the `isArchive` flag is being passed when creating ES Storage. In case of v1 it happens because v1 calls `factory.CreateArchiveWriter`, but in v2 all storages are initialized the same way so I don't think we actually call that function, we just call the main `CreateWriter`.

Comment (yurishkuro) [reply plugin/storage/es/spanstore/reader.go:127]: Booked [#6065](https://github.com/jaegertracing/jaeger/issues/6065)

**Comment by JaredTan95** [comment]
@mahadzaryab1 Thank you for your effort to keep this pr going

**Comment by mahadzaryab1** [comment]
@yurishkuro let me know if there any other items to address for this PR

File: plugin/storage/es/mappings/mapping.go [hunk 4]
```diff
@@ -69,17 +106,17 @@ func loadMapping(name string) string {
...
+func (mb *MappingBuilder) fixMapping(mapping string, options *IndexTemplateOptions) (string, error) {
 	tmpl, err := mb.TemplateBuilder.Parse(loadMapping(mapping))
 	if err != nil {
 		return "", err
 	}
 	writer := new(bytes.Buffer)
 
-	if mb.IndexPrefix != "" && !strings.HasSuffix(mb.IndexPrefix, "-") {
-		mb.IndexPrefix += "-"
+	if options.Prefix != "" && !strings.HasSuffix(options.Prefix, "-") {
```
Comment (yurishkuro) [inline plugin/storage/es/mappings/mapping.go:116]: Use separator const from config?

File: plugin/storage/es/options.go [hunk 5]
```diff
@@ -157,30 +166,30 @@ func addFlags(flagSet *flag.FlagSet, nsConfig *namespaceConfig) {
 		nsConfig.namespace+suffixTimeout,
 		nsConfig.Timeout,
 		"Timeout used for queries. A Timeout of zero means no timeout")
-	flagSet.Int64(
+	flagSet.Int(
 		nsConfig.namespace+suffixNumShards,
-		nsConfig.NumShards,
+		defaultIndexOptions.Shards,
```
Comment (yurishkuro) [inline plugin/storage/es/options.go:171]: This does not appear to be an equivalent change. Previously defaults were read from a struct passed as argument. Now they are read from a static struct. I think the intention of the original pattern was that you have a CreateDeafaultConfig function somewhere which populates stuff (including for v2) and cli flags don't invent their own defaults. Can we put it back to the original design?

Comment (yurishkuro) [reply plugin/storage/es/options.go:171]: Since cli flags don't allow per-index overrides for many options we can read them from the spans index settings.

File: plugin/storage/es/factory.go [hunk 6]
```diff
@@ -278,7 +274,7 @@ func createSpanWriter(
 		if err != nil {
 			return nil, err
 		}
-		if err := writer.CreateTemplates(spanMapping, serviceMapping, cfg.IndexPrefix); err != nil {
+		if err := writer.CreateTemplates(spanMapping, serviceMapping, cfg.Indices.Spans.Prefix); err != nil {
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply plugin/storage/es/factory.go:277]: in fact, if we extract IndexPrefix back to the top level, I don't think we'd need the other idea we discussed on the previous PR of having user-settable "defaults" applicable to each index

[the remark this answers is not in the document]
Comment (mahadzaryab1) [reply plugin/storage/es/factory.go:277]: @yurishkuro are you suggesting that we remove `Prefix` from `IndexOptions` and have it in `Configuration`?

[the remark this answers is not in the document]
Comment (yurishkuro) [reply plugin/storage/es/factory.go:277]: Yes (and call it IndexPrefix)

[the remark this answers is not in the document]
Comment (yurishkuro) [reply plugin/storage/es/factory.go:277]: btw, I would define it like this
```go
type IndexPrefix string

func (p IndexPrefix) Apply(indexName string) string { /* move logic of applying prefix here, to avoid repeating it */ }
```

File: plugin/storage/es/samplingstore/storage.go [hunk 7]
```diff
@@ -14,14 +14,14 @@ import (
...
+	"github.com/jaegertracing/jaeger/pkg/es/config"
 	"github.com/jaegertracing/jaeger/plugin/storage/es/samplingstore/dbmodel"
 )
 
 const (
-	samplingIndex        = "jaeger-sampling"
-	throughputType       = "throughput-sampling"
-	probabilitiesType    = "probabilities-sampling"
-	indexPrefixSeparator = "-"
+	samplingIndex     = "jaeger-sampling"
```
Comment (yurishkuro) [inline plugin/storage/es/samplingstore/storage.go:22]: to follow the convention from span store
```suggestion
	samplingIndexBaseName     = "jaeger-sampling"
```

File: plugin/storage/es/dependencystore/storage.go [hunk 8]
```diff
@@ -16,13 +16,13 @@ import (
...
+	"github.com/jaegertracing/jaeger/pkg/es/config"
 	"github.com/jaegertracing/jaeger/plugin/storage/es/dependencystore/dbmodel"
 )
 
 const (
-	dependencyType       = "dependencies"
-	dependencyIndex      = "jaeger-dependencies-"
-	indexPrefixSeparator = "-"
+	dependencyType  = "dependencies"
+	dependencyIndex = "jaeger-dependencies-"
```
Comment (yurishkuro) [inline plugin/storage/es/dependencystore/storage.go:25]:
```suggestion
	dependencyIndexBaseName = "jaeger-dependencies-"
```

File: plugin/storage/es/factory.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mahadzaryab1) [reply plugin/storage/es/factory.go:277]: @yurishkuro couple more questions regarding this: 
- how should we handle setting the `IndexPrefix` in `getMappingTemplateOptions`?
- how should we set the prefix in the span reader and writer? should they be different for the span index vs. service index?

[the remark this answers is not in the document]
Comment (yurishkuro) [reply plugin/storage/es/factory.go:277]: 1. What specifically? 
2. Before this PR there was just a single CLI flag `index-prefix`, so we never supported different prefix for spans and services.

File: plugin/storage/es/options.go [hunk 9]
```diff
@@ -157,30 +166,30 @@ func addFlags(flagSet *flag.FlagSet, nsConfig *namespaceConfig) {
 		nsConfig.namespace+suffixTimeout,
 		nsConfig.Timeout,
 		"Timeout used for queries. A Timeout of zero means no timeout")
-	flagSet.Int64(
+	flagSet.Int(
```
Comment (yurishkuro) [inline plugin/storage/es/options.go:169]: trying to remember why we changed this in the PR. I think since this PR started there were a few linter upgrades that created a lot of complaints with platform-specific `int`, so perhaps this change wasn't beneficial. With fixed-size ints you always know the range, regardless of the platform.

Comment (mahadzaryab1) [reply plugin/storage/es/options.go:169]: @yurishkuro do we want to change the configuration to hold `int64` as well? there's a type mismatch otherwise

**Comment by mahadzaryab1** [comment]
@yurishkuro do you have any idea why the v1 tests are failing?

File: cmd/es-rollover/app/init/flags.go [hunk 10]
```diff
@@ -23,12 +24,7 @@ const (
 // Config holds configuration for index cleaner binary.
 type Config struct {
 	app.Config
-	Shards                       int
-	Replicas                     int
-	PrioritySpanTemplate         int
-	PriorityServiceTemplate      int
-	PriorityDependenciesTemplate int
-	PrioritySamplingTemplate     int
+	cfg.Indices
```
Comment (mahadzaryab1) [inline cmd/es-rollover/app/init/flags.go:27]: @yurishkuro both `Indices` and `app.Config` have `IndexPrefix` in them and for the purpose of this package, we use the `IndexPrefix` from `app.Config`. Do you have any thoughts on what we should do here? It's a bit confusing at the moment.

Comment (mahadzaryab1) [reply cmd/es-rollover/app/init/flags.go:27]: Instead of embedding `cfg.Indices`, should we just add what is needed?

File: pkg/es/config/config.go [hunk 11]
```diff
@@ -33,48 +33,75 @@ import (
...
+}
+
+type IndexPrefix string
+
+func (p IndexPrefix) String() string {
+	return string(p)
+}
+
+func (p IndexPrefix) Apply(indexName string) string {
+	if p == "" {
```
Comment (yurishkuro) [inline pkg/es/config/config.go:65]: may want to do `ps := string(p)` to avoid repeating it multiple times.

File: pkg/es/config/config.go [hunk 12]
```diff
@@ -33,48 +33,75 @@ import (
...
+	IndexPrefix  IndexPrefix  `mapstructure:"index_prefix"`
+	Spans        IndexOptions `mapstructure:"spans"`
+	Services     IndexOptions `mapstructure:"services"`
+	Dependencies IndexOptions `mapstructure:"dependencies"`
+	Sampling     IndexOptions `mapstructure:"sampling"`
+}
+
+type IndexPrefix string
+
+func (p IndexPrefix) String() string {
```
Comment (yurishkuro) [inline pkg/es/config/config.go:60]: why is this needed?

File: plugin/storage/es/dependencystore/storage.go [hunk 13]
```diff
@@ -50,7 +50,7 @@ func NewDependencyStore(p Params) *DependencyStore {
 	return &DependencyStore{
 		client:                p.Client,
 		logger:                p.Logger,
-		dependencyIndexPrefix: prefixIndexName(p.IndexPrefix, dependencyIndex),
+		dependencyIndexPrefix: prefixIndexName(p.IndexPrefix, dependencyIndexBaseName),
```
Comment (yurishkuro) [inline plugin/storage/es/dependencystore/storage.go:53]:
```suggestion
		dependencyIndexPrefix: p.IndexPrefix.Apply(dependencyIndexBaseName),
```

File: plugin/storage/es/dependencystore/storage.go [hunk 14]
```diff
@@ -59,7 +59,7 @@ func NewDependencyStore(p Params) *DependencyStore {
 
 func prefixIndexName(prefix, index string) string {
```
Comment (yurishkuro) [inline plugin/storage/es/dependencystore/storage.go:60]: should not be needed, use Apply()

File: cmd/esmapping-generator/app/flags.go [hunk 15]
```diff
@@ -11,8 +11,8 @@ import (
 type Options struct {
 	Mapping       string
 	EsVersion     uint
-	Shards        int64
-	Replicas      int64
+	Shards        int
+	Replicas      int
```
Comment (yurishkuro) [inline cmd/esmapping-generator/app/flags.go:15]: can we undo these type changes?

File: pkg/es/config/config.go (hunk 12 above)
Comment (mahadzaryab1) [reply pkg/es/config/config.go:60]: just gives a convenient way to convert to a string instead of having to wrap in a `string()` everytime - i can remove it if you think we don't need it though

Comment (mahadzaryab1) [reply pkg/es/config/config.go]: removed

File: cmd/es-rollover/app/init/flags.go (hunk 10 above)
Comment (yurishkuro) [reply cmd/es-rollover/app/init/flags.go:27]: does `app.Config` actually need the prefix? Is it applicable to all sub-commands?

File: cmd/es-rollover/app/init/flags_test.go [hunk 16]
```diff
@@ -33,10 +33,10 @@ func TestBindFlags(t *testing.T) {
 	require.NoError(t, err)
 
 	c.InitFromViper(v)
-	assert.Equal(t, 8, c.Shards)
-	assert.Equal(t, 16, c.Replicas)
-	assert.Equal(t, 300, c.PrioritySpanTemplate)
-	assert.Equal(t, 301, c.PriorityServiceTemplate)
-	assert.Equal(t, 302, c.PriorityDependenciesTemplate)
-	assert.Equal(t, 303, c.PrioritySamplingTemplate)
+	assert.Equal(t, int64(8), c.Indices.Spans.Shards)
```
Comment (yurishkuro) [inline cmd/es-rollover/app/init/flags_test.go:36]: you can use `EqualValues` function without explicit casting.

File: pkg/es/config/config_test.go [hunk 17]
```diff
@@ -632,6 +619,41 @@ func TestValidate(t *testing.T) {
...
+			prefix:       "bye-",
+			name:         "hello",
+			expectedName: "bye-hello",
+		},
+		{
+			testName:     "no separator suffix",
+			prefix:       "bye",
+			name:         "hello",
+			expectedName: "bye-hello",
+		},
```
Comment (yurishkuro) [inline pkg/es/config/config_test.go:646]: you could also add a test with `name=""`

File: plugin/storage/es/mappings/mapping.go [hunk 18]
```diff
@@ -9,36 +9,70 @@ import (
...
+	EsVersion       uint
+	UseILM          bool
+	ILMPolicyName   string
+}
+
+// IndexTemplateOptions holds parameters required to render an elasticsearch index template
+type IndexTemplateOptions struct {
+	UseILM        bool
+	ILMPolicyName string
+	IndexPrefix   config.IndexPrefix
```
Comment (yurishkuro) [inline plugin/storage/es/mappings/mapping.go:33]: I think this should be a string. Semantically it's not the same as IndexPrefix, it's the full prefix+indexname, is it not?

File: plugin/storage/es/mappings/mapping.go [hunk 19]
```diff
@@ -9,36 +9,70 @@ import (
...
+	Shards        int64
+	Replicas      int64
+	Priority      int64
+}
+
+func (mb MappingBuilder) getMappingTemplateOptions(mapping string) *IndexTemplateOptions {
+	mappingOpts := &IndexTemplateOptions{}
+	mappingOpts.UseILM = mb.UseILM
+	mappingOpts.ILMPolicyName = mb.ILMPolicyName
+	mappingOpts.IndexPrefix = mb.Indices.IndexPrefix
```
Comment (yurishkuro) [inline plugin/storage/es/mappings/mapping.go:43]: can we NOT make this assignment? This field is being overwritten in L113

Comment (mahadzaryab1) [reply plugin/storage/es/mappings/mapping.go:43]: @yurishkuro L113 doesn't overwrite the field. It just adds a prefix to it if it doesn't exist.

File: plugin/storage/es/mappings/mapping.go (hunk 18 above)
Comment (mahadzaryab1) [reply plugin/storage/es/mappings/mapping.go:33]: @yurishkuro The reasoning for this was that we were doing this in fixMapping:
```go
	if options.IndexPrefix != "" && !strings.HasSuffix(options.IndexPrefix, config.IndexPrefixSeparator) {
		options.IndexPrefix += config.IndexPrefixSeparator
	}
```
which could be replaced with 
```go
	options.IndexPrefix = config.IndexPrefix(options.IndexPrefix.Apply(""))
```
so it seems to have similar semantics to the other usages. What do you think?

File: cmd/es-rollover/app/init/flags.go (hunk 10 above)
Comment (mahadzaryab1) [reply cmd/es-rollover/app/init/flags.go:27]: yeah - currently the `indexPrefix` field in `app.Config` is initialized in `InitFromViper` as 
```go
	c.IndexPrefix = v.GetString(indexPrefix)
	if c.IndexPrefix != "" {
		c.IndexPrefix += "-"
	}
```
and this is what is used in `cmd/es-rollover/app`

File: plugin/storage/es/mappings/mapping.go (hunk 18 above)
Comment (yurishkuro) [reply plugin/storage/es/mappings/mapping.go:33]: > `options.IndexPrefix = config.IndexPrefix(options.IndexPrefix.Apply(""))`

+1, but do you later do `options.IndexPrefix.Apply()` again? If yes then ok, it can be typed string.

File: plugin/storage/es/mappings/mapping.go (hunk 19 above)
Comment (yurishkuro) [reply plugin/storage/es/mappings/mapping.go:43]: what I mean is L113 does an unconditional assignment. You can replace that line with:

```
options.IndexPrefix = mb.Indices.IndexPrefix.Apply("")
```
and then `options.IndexPrefix` can be a plain string because you'd never call Apply on it again, you just pass it to the template (https://github.com/jaegertracing/jaeger/pull/6060/files#r1794418861).

File: plugin/storage/es/mappings/mapping.go [hunk 20]
```diff
@@ -9,36 +9,70 @@ import (
...
+type IndexTemplateOptions struct {
+	UseILM        bool
+	ILMPolicyName string
+	IndexPrefix   config.IndexPrefix
+	Shards        int64
+	Replicas      int64
+	Priority      int64
+}
+
+func (mb MappingBuilder) getMappingTemplateOptions(mapping string) *IndexTemplateOptions {
```
Comment (yurishkuro) [inline plugin/storage/es/mappings/mapping.go:39]:
```suggestion
func (mb MappingBuilder) getMappingTemplateOptions(mapping string) IndexTemplateOptions {
```

File: plugin/storage/es/mappings/mapping.go [hunk 21]
```diff
@@ -9,36 +9,70 @@ import (
...
-	ILMPolicyName                string
+	TemplateBuilder es.TemplateBuilder
+	Indices         config.Indices
+	EsVersion       uint
+	UseILM          bool
+	ILMPolicyName   string
+}
+
+// IndexTemplateOptions holds parameters required to render an elasticsearch index template
+type IndexTemplateOptions struct {
```
Comment (yurishkuro) [inline plugin/storage/es/mappings/mapping.go:30]: does this need to be public? can it be just `type params struct` ?

File: plugin/storage/es/mappings/mapping.go [hunk 22]
```diff
@@ -69,17 +103,15 @@ func loadMapping(name string) string {
 	return string(s)
 }
 
-func (mb *MappingBuilder) fixMapping(mapping string) (string, error) {
+func (mb *MappingBuilder) fixMapping(mapping string, options *IndexTemplateOptions) (string, error) {
```
Comment (yurishkuro) [inline plugin/storage/es/mappings/mapping.go:106]: can we rename this?
```suggestion
func (mb *MappingBuilder) renderMapping(mapping string, options *IndexTemplateOptions) (string, error) {
```

File: plugin/storage/es/mappings/mapping_test.go [hunk 23]
```diff
@@ -46,17 +47,30 @@ func TestMappingBuilderGetMapping(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.mapping, func(t *testing.T) {
+			indexTemOps := config.IndexOptions{
```
Comment (yurishkuro) [inline plugin/storage/es/mappings/mapping_test.go:50]:
```suggestion
			defaultOpts := func(p int) config.IndexOptions {
			  return config.IndexOptions{ ... }
			}
```

File: plugin/storage/es/mappings/mapping_test.go [hunk 24]
```diff
@@ -46,17 +47,30 @@ func TestMappingBuilderGetMapping(t *testing.T) {
...
-				PriorityServiceTemplate:      501,
-				PriorityDependenciesTemplate: 502,
-				EsVersion:                    tt.esVersion,
-				IndexPrefix:                  "test-",
-				UseILM:                       true,
-				ILMPolicyName:                "jaeger-test-policy",
+				TemplateBuilder: es.TextTemplateBuilder{},
+				Indices: config.Indices{
+					IndexPrefix:  "test-",
+					Spans:        indexTemOps,
```
Comment (yurishkuro) [inline plugin/storage/es/mappings/mapping_test.go:66]:
```suggestion
					Spans:        defaultOpts(500),
```

File: plugin/storage/es/spanstore/reader_test.go [hunk 25]
```diff
@@ -179,111 +178,116 @@ func TestSpanReaderIndices(t *testing.T) {
...
+	serviceIndexOpts := config.IndexOptions{DateLayout: serviceDataLayout}
+
 	testCases := []struct {
 		indices []string
 		params  SpanReaderParams
 	}{
 		{
 			params: SpanReaderParams{
-				IndexPrefix: "", Archive: false, SpanIndexDateLayout: spanDataLayout, ServiceIndexDateLayout: serviceDataLayout,
+				Archive: false, SpanIndex: spanIndexOpts, ServiceIndex: serviceIndexOpts,
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/reader_test.go:201]:
```suggestion
				Archive: false, 
				SpanIndex: spanIndexOpts, 
				ServiceIndex: serviceIndexOpts,
```

File: plugin/storage/es/spanstore/reader_test.go [hunk 26]
```diff
@@ -179,111 +178,116 @@ func TestSpanReaderIndices(t *testing.T) {
...
-				IndexPrefix: "foo:", Archive: true, UseReadWriteAliases: true,
+				SpanIndex: spanIndexOpts, ServiceIndex: serviceIndexOpts, IndexPrefix: "foo:", Archive: true, UseReadWriteAliases: true,
 			},
-			indices: []string{"foo:" + indexPrefixSeparator + spanIndex + archiveReadIndexSuffix, "foo:" + indexPrefixSeparator + serviceIndex + archiveReadIndexSuffix},
+			indices: []string{"foo:" + config.IndexPrefixSeparator + spanIndexBaseName + archiveReadIndexSuffix, "foo:" + config.IndexPrefixSeparator + serviceIndexBaseName + archiveReadIndexSuffix},
 		},
 		{
 			params: SpanReaderParams{
-				IndexPrefix: "", Archive: false, RemoteReadClusters: []string{"cluster_one", "cluster_two"}, SpanIndexDateLayout: spanDataLayout, ServiceIndexDateLayout: serviceDataLayout,
+				SpanIndex: spanIndexOpts, ServiceIndex: serviceIndexOpts, Archive: false, RemoteReadClusters: []string{"cluster_one", "cluster_two"},
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/reader_test.go:243]:
```suggestion
				SpanIndex: spanIndexOpts, 
				ServiceIndex: serviceIndexOpts, 
				Archive: false, 
				RemoteReadClusters: []string{"cluster_one", "cluster_two"},
```

File: plugin/storage/es/spanstore/writer.go [hunk 27]
```diff
@@ -76,51 +76,47 @@ func NewSpanWriter(p SpanWriterParams) *SpanWriter {
...
 
 // CreateTemplates creates index templates.
-func (s *SpanWriter) CreateTemplates(spanTemplate, serviceTemplate, indexPrefix string) error {
-	if indexPrefix != "" && !strings.HasSuffix(indexPrefix, "-") {
-		indexPrefix += "-"
-	}
-	_, err := s.client().CreateTemplate(indexPrefix + "jaeger-span").Body(spanTemplate).Do(context.Background())
+func (s *SpanWriter) CreateTemplates(spanTemplate, serviceTemplate string, indexPrefix cfg.IndexPrefix) error {
+	jaegerSpanId := indexPrefix.Apply("jaeger-span")
+	jaegerServiceId := indexPrefix.Apply("jaeger-service")
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/writer.go:86]:
```suggestion
	jaegerSpanIdx := indexPrefix.Apply("jaeger-span")
	jaegerServiceIdx := indexPrefix.Apply("jaeger-service")
```

File: plugin/storage/es/spanstore/writer_test.go [hunk 28]
```diff
@@ -40,7 +41,7 @@ func withSpanWriter(fn func(w *spanWriterTest)) {
 		client:    client,
 		logger:    logger,
 		logBuffer: logBuffer,
-		writer:    NewSpanWriter(SpanWriterParams{Client: func() es.Client { return client }, Logger: logger, MetricsFactory: metricsFactory, SpanIndexDateLayout: "2006-01-02", ServiceIndexDateLayout: "2006-01-02"}),
+		writer:    NewSpanWriter(SpanWriterParams{Client: func() es.Client { return client }, Logger: logger, MetricsFactory: metricsFactory, SpanIndex: config.IndexOptions{DateLayout: "2006-01-02"}, ServiceIndex: config.IndexOptions{DateLayout: "2006-01-02"}}),
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/writer_test.go:44]:
```suggestion
		writer:    NewSpanWriter(SpanWriterParams{
		     Client: func() es.Client { return client }, 
		     Logger: logger, 
		     MetricsFactory: metricsFactory, 
		     SpanIndex: config.IndexOptions{DateLayout: "2006-01-02"}, 
		     ServiceIndex: config.IndexOptions{DateLayout: "2006-01-02"}}),
```

File: plugin/storage/es/mappings/mapping.go (hunk 19 above)
Comment (mahadzaryab1) [reply plugin/storage/es/mappings/mapping.go:43]: ah okay i see! thanks so much for the explanation - this is fixed now

**Review by yurishkuro — APPROVED** [review]
